File: BACKEND/Typescript/src/controller/users.controller.ts

import { Request, Response } from 'express';
import { pool } from "../config/database/Postgres";


export const getAuthenticatedUser = async (req: Request, res: Response) => {
    try {
        const { id } = req.params;

        const usuario = await pool.query(
            `SELECT * FROM usuarios WHERE id = $1`,
            [id]
        );
        if (!usuario.rows[0]) {
            res.status(404).json({ message: 'User not found' });
            return;
        }
        res.status(200).json({
            user_id: id,
            nombres: usuario.rows[0].nombres,
            apellidos: usuario.rows[0].apellidos,
            email: usuario.rows[0].email,
            foto_perfil: usuario.rows[0].foto_perfil,
            password: usuario.rows[0].password_hash,
            rol: usuario.rows[0].rol,
            fecha_nacimiento: usuario.rows[0].fecha_nacimiento,
        });
        return;
    } catch (error: any) {
        res.status(500).json({ message: 'Internal server error', error: error.message });
        return;
    }
};

export const updateUser = async (req: Request, res: Response) => {
    try {
        const { id } = req.params;
        const { nombres, apellidos,email,fecha_nacimiento,rol } = req.body;

        const resultado = await pool.query(
            `UPDATE Usuarios 
             SET Nombres = $1, Apellidos = $2,Email = $3, Fecha_nacimiento = $4, Rol =$5
             WHERE id = $6
             RETURNING *`,
            [nombres, apellidos,email,fecha_nacimiento,rol,id]
        );
        if (!resultado.rows[0]) {
            res.status(404).json({ message: 'User not found' });
            return;
        }
        res.status(200).json({ message: 'Usuario actualizado con éxito' });

        return;
    }
    catch (error: any) {
        res.status(500).json({ message: 'Internal server error', error: error.message });
        return;
    }
};

export const listUserBooks = async (req: Request, res: Response) => {
    try {
        const { id } = req.params;
        
        const userQuery = await pool.query(
            `SELECT id FROM usuarios WHERE id = $1`,
            [id]
        );
        
        if (userQuery.rowCount === 0) {
            res.status(404).json({ error: 'Usuario no encontrado' });
            return;
        }

        const adquisicionesQuery = await pool.query(
            `SELECT 
                l.id, 
                l.nombre, 
                l.autor, 
                l.sinopsis, 
                l.portada_url, 
                l.pdf_url, 
                l.anio_publicacion AS publicacion
            FROM adquisiciones a
            JOIN libros l ON a.libro_id = l.id
            WHERE a.usuario_id = $1`,
            [id]
        );

        if (adquisicionesQuery.rowCount === 0) {
            res.status(404).json({ message: 'El usuario no ha adquirido ningún libro!' });
            return
        }
        
        res.status(200).json({
            user_id: id,
            total_books: adquisicionesQuery.rowCount,
            books: adquisicionesQuery.rows
        });
        return;
    } catch (error: any) {
        res.status(500).json({ message: 'Internal server error', error: error.message });
        return
    }
};

File: BACKEND/Python/src/controllers/users_controller.py
from flask import jsonify, request
from src.utils.db_connection import db
from sqlalchemy.orm import joinedload
from src.models.User import User
from src.models.Adquisicion import Adquisicion


def update_controller(_id, data):
    try:
        user = User.query.get(_id)
        if not user:
            return jsonify(error="Usuario no encontrado"), 404
        
        # Verificar solo si se intenta cambiar el email
        if 'email' in data:
            nuevo_email = data['email'].lower()
            if nuevo_email != user.email.lower():
                # Buscar si otro usuario ya tiene el nuevo email
                existing_user = User.query.filter(
                    User.email == nuevo_email,
                    User.id != user.id  # Excluir al usuario actual
                ).first()
                if existing_user:
                    return jsonify(error="El email ya está registrado en otro usuario"), 400
                user.email = nuevo_email  # Actualizar solo si pasa la validación

        # Actualizar otros campos si existen en 'data' y son diferentes
        if 'nombres' in data and data['nombres'] != user.nombres:
            user.nombres = data['nombres']
        if 'apellidos' in data and data['apellidos'] != user.apellidos:
            user.apellidos = data['apellidos']
        if 'fecha_nacimiento' in data and data['fecha_nacimiento'] != user.fecha_nacimiento:
            user.fecha_nacimiento = data['fecha_nacimiento']
        if 'rol' in data and data['rol'] != user.rol:
            user.rol = data['rol']

        db.session.commit()
        return jsonify(message="Usuario actualizado con éxito"), 200
    
    except Exception as e:
        db.session.rollback()  # Revertir cambios en caso de error
        return jsonify(error=str(e)), 500
    
def user_books_controller(_id):
    try:
        user = db.session.query(User)\
        .options(joinedload(User.adquisiciones).joinedload(Adquisicion.libro))\
        .filter(User.id == _id)\
        .first()

        if not user:
            return jsonify(error = "Usuario no encontrado"), 404
        
        book_list = [
            {
                "id": adquisicion.libro.id,
                "nombre": adquisicion.libro.nombre,
                "autor": adquisicion.libro.autor,
                "sinopsis": adquisicion.libro.sinopsis,
                "portada_url": adquisicion.libro.portada_url,
                "pdf_url": adquisicion.libro.pdf_url,
                "publicacion": adquisicion.libro.anio_publicacion
            }
            for adquisicion in user.adquisiciones
        ]

        if len(book_list) == 0:
            return jsonify(message = "El usuario no ha aquirido ningun libro!"), 404

        return jsonify(user_id= _id, total_books = len(book_list), books = book_list), 200
    except Exception as e:
        return jsonify(error = str(e)), 500
    
def get_user_controller(id):
    try:
        user = User.query.get(id)
        if not user:
            return jsonify(error="Usuario no encontrado"), 404
        
        return jsonify(
            id = user.id,
            nombres = user.nombres,
            apellidos = user.apellidos,
            email = user.email,
            password_hash = user.password_hash,
            foto_perfil_url = user.foto_perfil_url,
            fecha_nacimiento = user.fecha_nacimiento,
            rol = user.rol
        ), 200

    except Exception as e:
        return jsonify(error = str(e)), 500